### PR TITLE
Generate and verify commitments

### PR DESCRIPTION
**Pull request by sune-aion**
Adds generation and verification of encryption commitments.

As part of the discussion on how to turn aion_crypto into a tested, standalone, high quality library this PRs proposes to start splitting things into smaller separate modules, and using static typing (typescript).

As the underlying crypto implementation (SJCL) uses untyped JS classes, this PR introduces typed wrapper classes for Point and Bignum.

* Written in typescript, and using a typescript wrapper of aion_crypto lib.
* Added unit tests for generator & verifier functions.
* Validated against commitment generator implementation in the backend

### CODE REVIEW

File: lib/av_client/crypto/pedersen_commitment.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    const target = [baseGeneratorPrefix(), index].join(',');
+    let x = hashToBignum(target).mod(secp256k1_curve_prime);
+
+    /*eslint no-constant-condition: ["error", { "checkLoops": false }]*/
+    while(true) {
+      try {
+        return pointFromX(x);
+      }
+      catch(err) {
+        // Skip to next iteration
```
Comment (mzedeler) [inline lib/av_client/crypto/pedersen_commitment.ts:29]: Please explain why this is ok.

File: lib/av_client/crypto/pedersen_commitment.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  pointFromX,
+  pointToHex,
+} from "./util";
+
+class PedersenCommitment {
+  static computeGenerator(index) {
+    const baseGeneratorPrefix = () => pointToHex(new Point(Curve.G));
+    const secp256k1_curve_prime = new Bignum(Curve.field.modulus);
+
+    const target = [baseGeneratorPrefix(), index].join(',');
```
Comment (mzedeler) [inline lib/av_client/crypto/pedersen_commitment.ts:20]: Just a question: what is `target`? I can see it contains a string with a comma (and probably numbers on each side), but it is unclear why this is a good idea.

File: lib/av_client/crypto/pedersen_commitment.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * @description Generates an encryption commitment
+ * @param messages An array of hex strings
+ * @param options Optional options object. Allows caller to specify randomizer
+ * @returns Commitment point and randomizer, both as hex.
+ */
+const generatePedersenCommitment = (messages: string[], options?: { randomizer: string }) => {
+  const bnMessages: Bignum[] = messages.map(m => bignumFromHex(m));
+
+  const randomizer = options && options.randomizer ?
```
Comment (mzedeler) [inline lib/av_client/crypto/pedersen_commitment.ts:63]: `options && options.randomizer` => `options?.randomizer`.

File: lib/av_client/crypto/pedersen_commitment.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    new Bignum(options.randomizer) : generateRandomBignum();
+
+  const commitment = PedersenCommitment.generate(bnMessages, randomizer);
+
+  return {
+    commitment: pointToHex(commitment),
+    randomizer: bignumToHex(randomizer),
```
Comment (mzedeler) [inline lib/av_client/crypto/pedersen_commitment.ts:70]: Is `randomizer` the right name? It is very generic and sounds more like a verb (indicating function or object) than a noun.

File: lib/av_client/crypto/util.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+export const bignumFromHex = (hex: string): Bignum => new Bignum(sjcl.bn.fromBits(sjcl.codec.hex.toBits(hex)));
+export const bignumToHex = (bignum: Bignum): string => sjcl.codec.hex.fromBits(bignum.toBits());
+
+export const hashToBignum = (hash: BitArray): Bignum => new Bignum(crypto.hashToBn(hash));
+
+// Other
+// --------------------------
+export const generateRandomBignum = () => new Bignum(crypto.randomBN());
+
+export const pointFromX = (x: Bignum): Point => {
```
Comment (mzedeler) [inline lib/av_client/crypto/util.ts:27]: This should be covered by tests.

File: lib/av_client/crypto/util.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  return new Point(pointFromBits(encodedPoint));
+}
+
+export const addPoints = (a: Point, b: Point): Point => {
+  return new Point(crypto.addPoints(a.toEccPoint(), b.toEccPoint()));
+}
+
+// Types
+// --------------------------
+export class Bignum {
```
Comment (mzedeler) [inline lib/av_client/crypto/util.ts:42]: I would move this to a separate file and add tests.

File: lib/av_client/crypto/util.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+  mod = (operand: Bignum): Bignum => new Bignum(this.bn.mod(operand.bn));
+  add = (operand: Bignum): Bignum => new Bignum(this.bn.add(operand.bn))
+
+  toBits = () => this.bn.toBits();
+  toBn = () => this.bn;
+}
+
+
+export class Point {
```
Comment (mzedeler) [inline lib/av_client/crypto/util.ts:60]: Same as above.

File: test/crypto/commitment.test.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
+import { expect } from "chai";
+import { generatePedersenCommitment, isValidPedersenCommitment } from "../../lib/av_client/crypto/pedersen_commitment";
+
+describe("Pedersen Commitments", () => {
+  it("generate commitment from single message", () => {
```
Comment (mzedeler) [inline test/crypto/commitment.test.ts:5]: Maybe add some tests that document how errors and edge cases are handled?
What if `messages === []`?
Can `randomizer` be too long or short? Should it be limited to certain chars or are all characters allowed?

File: lib/av_client/crypto/pedersen_commitment.ts (hunk 1 above)
Comment (sune-aion) [reply lib/av_client/crypto/pedersen_commitment.ts:29]: Same as above. My understanding is that there's a 50% chance that a point will be found in a given iteration. SJCL will throw an error if an x-value does not correspond to a valid point on the curve.

Looking at it again, I see that there is of course the possibility that "some other" error occurs, which cannot be alleviated by incrementing the x-value -> Possibly leading to an infinite loop.

I will allow it to only continue for that specific error - and rethrow any other error.

File: lib/av_client/crypto/pedersen_commitment.ts (hunk 4 above)
Comment (sune-aion) [reply lib/av_client/crypto/pedersen_commitment.ts:70]: I agree. The trade-off is that this is the name used in our (academic) documentation.

Comment (sune-aion) [reply lib/av_client/crypto/pedersen_commitment.ts:70]: My understanding is that this corresponds to the cryptographic concept referred to as "a/the randomizer".

File: lib/av_client/crypto/util.ts (hunk 5 above)
Comment (sune-aion) [reply lib/av_client/crypto/util.ts:27]: It is tested by `commitment.test.ts`. There's a ~99% coverage of the code in the PR.

File: lib/av_client/crypto/pedersen_commitment.ts (hunk 1 above)
Comment (mzedeler) [reply lib/av_client/crypto/pedersen_commitment.ts]: Yes, but even if we use a catch all, please explain why the `catch` is a no-op.

File: lib/av_client/crypto/pedersen_commitment.ts (hunk 4 above)
Comment (mzedeler) [reply lib/av_client/crypto/pedersen_commitment.ts:75]: Ok

File: lib/av_client/crypto/util.ts (hunk 5 above)
Comment (mzedeler) [reply lib/av_client/crypto/util.ts]: I must have overlooked it. Thanks.

File: lib/av_client/crypto/pedersen_commitment.ts (hunk 1 above)
Comment (sune-aion) [reply lib/av_client/crypto/pedersen_commitment.ts]: Got it. See this commit 4c27e4a.

File: lib/av_client/crypto/util.ts (hunk 6 above)
Comment (sune-aion) [reply lib/av_client/crypto/util.ts]: Done

File: test/crypto/commitment.test.ts (hunk 8 above)
Comment (sune-aion) [reply test/crypto/commitment.test.ts:5]: Good point. I added some more boundary-oriented tests.

File: lib/av_client/crypto/pedersen_commitment.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      x = x.add(new Bignum(1)).mod(secp256k1_curve_prime);
+    }
+  }
+
+  static verify(commitment: Point, messages, randomizer: Bignum) {
+    return commitment.equals(this.generate(messages, randomizer))
+  }
+
+  static generate(messages: Bignum[], randomizer: Bignum): Point {
+    let commitment = new Point(Curve.G).mult(randomizer);
```
[the remark this answers is not in the document]
Comment (sune-aion) [reply lib/av_client/crypto/pedersen_commitment.ts]: Done

**Review by oteyatosys — APPROVED** [review]
